clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1465,13 +1465,17 @@ public boolean isDone() {
     private class AppendCallbacks<K, V> implements RecordAccumulator.AppendCallbacks {
         private final Callback userCallback;
         private final ProducerInterceptors<K, V> interceptors;
-        private final ProducerRecord<K, V> record;
+        private ProducerRecord<K, V> record;
+        private final String topic;
         protected int partition = RecordMetadata.UNKNOWN_PARTITION;
 
         private AppendCallbacks(Callback userCallback, ProducerInterceptors<K, V> interceptors, ProducerRecord<K, V> record) {
             this.userCallback = userCallback;
             this.interceptors = interceptors;
             this.record = record;
+            // Note a record would be null only if the client application has a bug, but we don't want to
+            // have NPE here, because the interceptors would not be notified (see .doSend).
+            topic = record != null ? record.topic() : null;
```

Comment:
Can you elaborate on this? What kind of application bug would surface itself in a silent way like this?

Reply:
https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala#L1041 has a test, which effectively codifies the contract.  I would agree that it's weird to have contract about null handling, but at this point I'd rather preserve whatever behavior is codified.

Reply:
I don't think this answered my question. What application bug would result in this?

Reply:
If a user passes in a null record in send(), we will be throwing a NullPointException somewhere. So, we probably could just throw an exception early in that case without going through the callback and fix the test accordingly. We probably could do that in a separate PR in trunk.

Reply:
I looked at the test and it seems to check that an exception is thrown? As @junrao said, this can be done by validating what `send` receives instead of polluting the whole codebase.

Reply:
It checks that the exception is thrown and then it checks that interceptors are called.  Probably the test is just sloppy and could use a different error condition. KAFKA-14086